### PR TITLE
Drone Syncing

### PR DESCRIPTION
**Pull request by sp00ktober**
Adds syncing of Mecha and BattleBase construction/repair drones. As it still contains at least one bug and a lot of debug prints i'll set it as draft for now.

One issue i encounter is a deadlock in `DroneManager` `PlayerDroneBuildingPlans` and `PendingBuildRequests` leading to prebuilds that cant be completed by drones.
This can be fixed by instant building broken parts or sometimes by building more stuff as some drones eventually will build the broken structures fixing the issue.

Im not yet sure whats causing it..

### CODE REVIEW

File: NebulaNetwork/PacketProcessors/Factory/BattleBase/NewBattleBaseDroneOrderProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+using NebulaWorld.Player;
+
+namespace NebulaNetwork.PacketProcessors.Factory.BattleBase
+{
+    [RegisterPacketProcessor]
+    internal class NewBattleBaseDroneOrderProcessor : PacketProcessor<NewBattleBaseDroneOrderPacket>
+    {
+        protected override void ProcessPacket(NewBattleBaseDroneOrderPacket packet, NebulaConnection conn)
+        {
+            if (IsHost)
```
Comment (PhantomGamers) [inline NebulaNetwork/PacketProcessors/Factory/BattleBase/NewBattleBaseDroneOrderProcessor.cs:19]: return here to reduce nesting

File: NebulaNetwork/PacketProcessors/Factory/BattleBase/NewBattleBaseDroneOrderProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            }
+            else
+            {
+                if (packet.PlanetId != GameMain.mainPlayer.planetId)
+                {
+                    return;
+                }
+
+                PlanetFactory factory = GameMain.galaxy.PlanetById(packet.PlanetId)?.factory;
+                if (factory != null)
```
Comment (PhantomGamers) [inline NebulaNetwork/PacketProcessors/Factory/BattleBase/NewBattleBaseDroneOrderProcessor.cs:31]: invert if to reduce nesting (return if factory is null)

File: NebulaNetwork/PacketProcessors/Factory/BattleBase/NewBattleBaseDroneOrderProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                {
+                    BattleBaseComponent battleBaseComponent = factory.defenseSystem.battleBases.buffer[packet.Owner];
+                    double droneEjectEnergy = Configs.freeMode.droneEjectEnergy;
+
+                    DroneComponent[] dronesBuffer = factory.constructionSystem.drones.buffer;
+                    int cursor = factory.constructionSystem.drones.cursor;
+                    for (int i = 1; i < cursor; i++)
+                    {
+                        ref DroneComponent drone = ref dronesBuffer[i];
+                        if (drone.id == i && drone.owner == packet.Owner && drone.stage == 0) // checking for stage 0 is important to not reuse the same drone again and again.
```
Comment (PhantomGamers) [inline NebulaNetwork/PacketProcessors/Factory/BattleBase/NewBattleBaseDroneOrderProcessor.cs:41]: invert if and return early to reduce  nesting

**Review by PhantomGamers — COMMENTED** [review]
There are some other nesting cases here that I didn't comment on because it's hard to view in the github editor so I will have to check it locally later but I think we should try to avoid nesting as much as possible as it does negatively impact code readability.

**Comment by PhantomGamers** [comment]
I posted this PR to add onto this https://github.com/sp00ktober/nebula/pull/2

**Review by starfi5h — APPROVED** [review]

**Pull request merged**